File: trunk/IvanGame/Sprite.cpp
#include "Sprite.h"
#include "Game.h"
#include <iostream>
#include <GL/gl.h>

using std::cout;
using std::endl;
Sprite::Sprite(SDL_Surface *screen, std::string name, Actor actor) :
mName(name),
	mDrawn(0),
	mLastUpdate(0),
	mActor(actor),
	mScreen(screen)
{
	//add to current level
	Game::game()->getCurrentLevel()->addSprite((this));

	if(mActor.mAnimations[mActor.mCurrentAnimation] -> mBuilt)
	{
		if (mActor.mAnimations[mActor.mCurrentAnimation]->mNumFrames > 1) mAnimating = 1;
	}
}

void Sprite::draw()
{
	//Frame so we don't have to find it so often
	SpriteFrame* frame = getAnimation();
	if(mAnimating == 1) {
		if(mLastUpdate+frame->pause*mSpeed<SDL_GetTicks()) {
			//obtain current peg 
			Point2D ppos = frame->animationPeg;
			mActor.mFrame++;
			if(mActor.mFrame > mActor.mAnimations[mActor.mCurrentAnimation]->mNumFrames-1)
				mActor.mFrame=0;
			//update frame so we don't need to worry
			frame = &mActor.mAnimations[mActor.mCurrentAnimation]->mFrames[mActor.mFrame];
			//obtain next peg
			Point2D npos = frame->animationPeg;
			//move current position to difference of two
			mPos.add(ppos - npos);
			mLastUpdate = SDL_GetTicks();
		}
	}
	if(mVisible == true){
		glEnable(GL_BLEND);
		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
		glColor4f(1, 1, 1, mTransparency);
		glLoadIdentity();
		glTranslated(mPos.x,mPos.y,0);//move to position
		glTranslated(frame->animationPeg.x,frame->animationPeg.y,0);//move back to where it was
		glRotatef(mAngle,0.0f,0.0f,1.0f);//rotate shape
		glTranslated(-frame->animationPeg.x,-frame->animationPeg.y,0);//place animation peg on origin
		glBindTexture(GL_TEXTURE_2D, frame->image);
		glBegin( GL_QUADS );
		// Top-left vertex (corner)
		glTexCoord2f( 0, 0 );
		glVertex3d( 0, 0, 0 );

		// Bottom-left vertex (corner)
		glTexCoord2f( 1, 0 );
		glVertex3d( frame->width, 0, 0 );

		// Bottom-right vertex (corner)
		glTexCoord2f( 1, 1 );
		glVertex3d( frame->width, frame->height, 0 );

		// Top-right vertex (corner)
		glTexCoord2f( 0, 1 );
		glVertex3d( 0, frame->height, 0 );
		glEnd();
		glDisable(GL_BLEND);
		glLoadIdentity();
	}

}

vector<Collision*>& Sprite::getCollisionData(){
	return mActor.mAnimations[mActor.mCurrentAnimation]->mFrames[mActor.mFrame].collisionData;
}


void Sprite::drawCollisions(){
	//get the frame for readability
	SpriteFrame frame = mActor.mAnimations[mActor.mCurrentAnimation]->mFrames[mActor.mFrame];
	//center the location
	Point2D pt = mPos + frame.animationPeg;
	WorldObject::drawCollisions(frame.collisionData, pt);
}

Sprite* Sprite::collisionWithSprite(string name){
	//get the frame for readability
	SpriteFrame* frame = getAnimation();
	//get the first sprite with this name
	Sprite* s= Game::game()->getCurrentLevel()->findSpriteByName(name);
	if(s==NULL)
		return NULL;
	for(unsigned int i=0; i <  frame->collisionData.size(); i++)
		if(frame->collisionData[i]->checkCollisions(s->getCollisionData(), s->mPos + s->getAnimation()->animationPeg, mPos + frame->animationPeg))//if there is a collision
			return s;
	return NULL;//if there aren't collisions
}

File: trunk/BodilyInfection/BodilyInfection/Sprite.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.Xna.Framework;

namespace BodilyInfection
{
    class Sprite : WorldObject
    {
        public Sprite(string name, Actor actor)
        {
            mName = name;
            mDrawn = false;
            mActor = actor;
            mLastUpdate = new GameTime();
            //add to current level
            This.Game.CurrentLevel.AddSprite(this);

            if (mActor.Animations[mActor.CurrentAnimation].Built)
            {
                if (mActor.Animations[mActor.CurrentAnimation].NumFrames > 1) mAnimating = true;
            }
        }

        #region Properties
        /// <summary>
        ///     changes to the specified frame of the animation beginning at 0
        /// </summary>
        public int Frame { get; set; }

        /// <summary>
        ///     the sprite's speed
        /// </summary>
        public float Speed { get; set; }

        /// <summary>
        ///     gets the sprite's name
        /// </summary>
        public string Name { get { return mName; } }

        /// <summary>
        ///     changes to the specified animation beginning at 0
        /// </summary>
        public SpriteFrame GetAnimation()
        {
            return mActor.Animations[mActor.CurrentAnimation].Frames[mActor.Frame];
        }

        #endregion Properties


        #region Methods
        /// <summary>
        /// changes to the specified animation beginning at 0.
        /// </summary>
        /// <param name="animation">The animation to select (begins at 0)</param>
        public void SetAnimation(int animation) { mActor.CurrentAnimation = animation; }

        /// <summary>
        /// Pauses or resumes an animation.
        /// </summary>
        public void ToggleAnim() { mAnimating = !mAnimating; }

        /// <summary>
        /// Causes the animation to play.
        /// </summary>
        public void StartAnim() { mAnimating = true; }

        /// <summary>
        /// Causes the animation to stop.
        /// </summary>
        public void StopAnim() { mAnimating = false; }

        /// <summary>
        ///  Resets the Sprite's animation to the first frame.
        /// </summary>
        public void Rewind() { mActor.Frame = 0; }

        /// <summary>
        /// Draw the Scene
        /// </summary>
        /// <param name="gameTime">Game time as given by the game class</param>
        public override void Draw(Microsoft.Xna.Framework.GameTime gameTime)
        {
            //Frame so we don't have to find it so often
            SpriteFrame frame = GetAnimation();
            if (mAnimating == true)
            {
                if (mLastUpdate.TotalGameTime.Milliseconds + frame.Pause * mSpeed < gameTime.TotalGameTime.Milliseconds)
                {
                    //obtain current peg 
                    Vector2 ppos = frame.AnimationPeg;
                    mActor.Frame++;
                    if (mActor.Frame > mActor.Animations[mActor.CurrentAnimation].NumFrames - 1)
                        mActor.Frame = 0;
                    //update frame so we don't need to worry
                    frame = mActor.Animations[mActor.CurrentAnimation].Frames[mActor.Frame];
                    //obtain next peg
                    Vector2 npos = frame.AnimationPeg;
                    //move current position to difference of two
                    Pos += (ppos - npos);
                    mLastUpdate = new GameTime(gameTime.TotalGameTime,gameTime.ElapsedGameTime);
                }
            }
            if (mVisible == true)
            {
                This.Game.spriteBatch.Draw(mActor.Animations[mActor.CurrentAnimation].Frames[mActor.Frame].image, Pos, Color.White);
            }
        }

        /// <summary>
        /// Draws Collision data for the Object (from outside)
        /// </summary>
        public void DrawCollisions()
        {
            //get the frame for readability
            SpriteFrame frame = mActor.Animations[mActor.CurrentAnimation].Frames[mActor.Frame];
            //center the location
            Vector2 pt = Pos + frame.AnimationPeg;
            drawCollisions(frame.CollisionData, pt);
        }

        /// <summary>
        /// checks for collision with sprite of a given name.
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public Sprite CollisionWithSprite(string name)
        {
            //        //get the frame for readability
            //SpriteFrame frame = GetAnimation();
            ////get the first sprite with this name
            //Sprite s= This.Game.getCurrentLevel().findSpriteByName(name);
            //if(s==null)
            //    return null;
            //for(int i=0; i <  frame.CollisionData.Count; i++)
            //    if(frame.CollisionData[i].checkCollisions(s.getCollisionData(), s.Pos + s.GetAnimation().AnimationPeg, Pos + frame.AnimationPeg))//if there is a collision
            //        return s;
            return null;//if there aren't collisions
        }
        /// <summary>
        /// Returns collision data
        /// </summary>
        /// <returns></returns>
        //public vector<Collision> getCollisionData();
        #endregion Methods

        #region Variables
        /// <summary>
        /// Sprite's name
        /// </summary>
        private string mName;

        /// <summary>
        /// Tells whether to animate or not
        /// </summary>
        private bool mAnimating;

        /// <summary>
        /// Tells if the object has been drawn the first time
        /// </summary>
        private bool mDrawn;

        /// <summary>
        /// Movement speed of the Sprite.
        /// </summary>
        private float mSpeed;

        /// <summary>
        /// Number that indicates when the Sprite has last updated. Overflows in about 24 days so no worries.
        /// </summary>
        private Microsoft.Xna.Framework.GameTime mLastUpdate;

        /// <summary>
        /// This Sprite's Actor.
        /// </summary>
        private Actor mActor;

        /// <summary>
        /// Sprite's Behavior
        /// </summary>
        private Behavior mBehavior;
        #endregion Variables
    }
}
